### PR TITLE
Wait for custom conditions on objects

### PR DESCRIPTION
**Pull request by fcanovai**
### Description

Allow users to define conditions on which to wait, overriding the default waiters behaviour. This can be used to check readiness also of custom resources, which were previously skipped.

### Fixes

The second commit includes a fix for a race condition. The `verifyCondition` function could run before the object being created is actually there, meaning that the `List` in https://github.com/fcanovai/kube-burner/blob/1df0567dce5bd9618a8818c0bf4275bf59232321/pkg/burner/waiters.go#L237 could return an empty list, and the function can exit without actually waiting. The change in the waitgroup allows the creation of all the resources before checking for their existence.

### CODE REVIEW

**Comment by rsevilla87** [comment]
Hey @fcanovai, first off, thanks for your contribution. This PR is a great addition.

With respect to the the bug you try to address at https://github.com/cloud-bulldozer/kube-burner/pull/308/commits/1df0567dce5bd9618a8818c0bf4275bf59232321, I have tested it locally with good results, the bug seems to be fixed and now kube-burner waits for all objects to be created before triggering any "waiter" function, however it comes with some trade-offs like for example, QPS and BURST rates not being met in scenarios when the number of iterations is 1 and all objects are create within the same namespace

This can be observed in the node-density workload implemented in the ocp wrapper:

```console
./bin/amd64/kube-burner ocp node-density --pods-per-node=75 --gc=false  --alerting=false --log-level=debug
time="2023-05-08 11:55:31" level=debug msg="Prometheus endpoint: <truncated>
time="2023-05-08 11:55:31" level=debug msg="Bearer token: <truncated>
time="2023-05-08 11:55:33" level=info msg="Listed nodes after using selector node-role.kubernetes.io/worker=,node-role.kubernetes.io/infra!=,node-role.kubernetes.io/workload!=: 3" file="discovery.go:117"
time="2023-05-08 11:55:39" level=debug msg="Current running pod count: 167" file="discovery.go:135"
time="2023-05-08 11:55:39" level=info msg="File node-density.yml available in the current directory, using it" file="helpers.go:212"
time="2023-05-08 11:55:39" level=info msg="👽 Initializing prometheus client with URL: https://prometheus-k8s-openshift-monitoring.apps.cloud-bulldozer-ci.perfscale.devcluster.openshift.com" file="prometheus.go:55"
time="2023-05-08 11:55:40" level=debug msg="Prometheus endpoint verified" file="prometheus.go:87"
time="2023-05-08 11:55:40" level=info msg="🔥 Starting kube-burner (latest@4034cbedf4f841b841b576ddf01047dd1eb3b026) with UUID a789fa04-1cd5-4abd-826d-0ef11b8b3117" file="job.go:86"
time="2023-05-08 11:55:40" level=info msg="📈 Creating measurement factory" file="factory.go:52"
time="2023-05-08 11:55:40" level=info msg="Registered measurement: podLatency" file="factory.go:85"
time="2023-05-08 11:55:40" level=debug msg="Preparing create job: node-density" file="create.go:49"
time="2023-05-08 11:55:40" level=debug msg="Rendering template: pod.yml" file="create.go:60"
time="2023-05-08 11:55:40" level=info msg="Job node-density: 58 iterations with 1 Pod replicas" file="create.go:91"
time="2023-05-08 11:55:40" level=info msg="QPS: 20" file="job.go:96"
time="2023-05-08 11:55:40" level=info msg="Burst: 20" file="job.go:97"
time="2023-05-08 11:55:40" level=info msg="Triggering job: node-density" file="job.go:111"
time="2023-05-08 11:55:40" level=info msg="Creating Pod latency watcher for node-density" file="pod_latency.go:137"
time="2023-05-08 11:55:45" level=info msg="Running job node-density" file="create.go:105"
time="2023-05-08 11:55:45" level=debug msg="Created namespace: node-density" file="namespaces.go:52"
time="2023-05-08 11:55:45" level=debug msg="Creating object replicas from iteration 1" file="create.go:124"
time="2023-05-08 11:55:45" level=debug msg="Created Pod/node-density-1 in namespace node-density" file="create.go:245"
time="2023-05-08 11:55:45" level=debug msg="Creating object replicas from iteration 2" file="create.go:124"
time="2023-05-08 11:55:46" level=debug msg="Created Pod/node-density-2 in namespace node-density" file="create.go:245"
time="2023-05-08 11:55:46" level=debug msg="Creating object replicas from iteration 3" file="create.go:124"
time="2023-05-08 11:55:46" level=debug msg="Created Pod/node-density-3 in namespace node-density" file="create.go:245"
time="2023-05-08 11:55:46" level=debug msg="Creating object replicas from iteration 4" file="create.go:124"
time="2023-05-08 11:55:46" level=debug msg="Created Pod/node-density-4 in namespace node-density" file="create.go:245"
time="2023-05-08 11:55:46" level=debug msg="Creating object replicas from iteration 5" file="create.go:124"
time="2023-05-08 11:55:47" level=debug msg="Created Pod/node-density-5 in namespace node-density" file="create.go:245"
time="2023-05-08 11:55:47" level=info msg="5/58 iterations completed" file="create.go:121"
time="2023-05-08 11:55:47" level=debug msg="Creating object replicas from iteration 6" file="create.go:124"
time="2023-05-08 11:55:47" level=debug msg="Created Pod/node-density-6 in namespace node-density" file="create.go:245"
time="2023-05-08 11:55:47" level=debug msg="Creating object replicas from iteration 7" file="create.go:124"
time="2023-05-08 11:55:47" level=debug msg="Created Pod/node-density-7 in namespace node-density" file="create.go:245"
time="2023-05-08 11:55:47" level=debug msg="Creating object replicas from iteration 8" file="create.go:124"
time="2023-05-08 11:55:48" level=debug msg="Created Pod/node-density-8 in namespace node-density" file="create.go:245"
time="2023-05-08 11:55:48" level=debug msg="Creating object replicas from iteration 9" file="create.go:124"
time="2023-05-08 11:55:48" level=debug msg="Created Pod/node-density-9 in namespace node-density" file="create.go:245"
time="2023-05-08 11:55:48" level=debug msg="Creating object replicas from iteration 10" file="create.go:124"
time="2023-05-08 11:55:48" level=debug msg="Created Pod/node-density-10 in namespace node-density" file="create.go:245" 
time="2023-05-08 11:55:48" level=info msg="10/58 iterations completed" file="create.go:121"
time="2023-05-08 11:55:48" level=debug msg="Creating object replicas from iteration 11" file="create.go:124"
time="2023-05-08 11:55:48" level=debug msg="Created Pod/node-density-11 in namespace node-density" file="create.go:245" 
time="2023-05-08 11:55:48" level=debug msg="Creating object replicas from iteration 12" file="create.go:124"

```

Note that in the previous output using a binary compiled including your fix, the object creation rate doesn't match with the QPS/Burst values 20.


The problem of your implementation is that kube-burner will block the execution per each iteration, until all objects are created.

The diff below contains another approach to fix the original bug you reported, which is easier and does not create this issue. 

```diff
 $ git diff
diff --git a/pkg/burner/create.go b/pkg/burner/create.go
index f3aea42..dab8c51 100644
--- a/pkg/burner/create.go
+++ b/pkg/burner/create.go
@@ -135,6 +135,7 @@ func (ex *Executor) RunCreateJob(iterationStart, iterationEnd int) {
                }
                if ex.Config.PodWait {
                        log.Infof("Waiting up to %s for actions to be completed in namespace", ex.Config.MaxWaitTimeout, ns)
+                       wg.Wait()
                        ex.waitForObjects(ns)
                }
                if ex.Config.JobIterationDelay > 0 {
```

**Comment by rsevilla87** [comment]
> Thanks for the quick review! The second approach works for me as well. I can remove the second commit from the PR or change it to your suggestion. What approach would you prefer?

Hey @fcanovai I think you can remove/modify your 2nd commit and replace it with my suggestion if you don't mind

**Comment by fcanovai** [comment]
I don't mind, go ahead :)

File: pkg/config/types.go [hunk 1]
```diff
@@ -138,3 +140,8 @@ type Job struct {
 	// Churn delay between sets
 	ChurnDelay time.Duration `yaml:"churnDelay" json:"churnDelay,omitempty"`
 }
+
+type WaitOptions struct {
+	// ForCondition wait for this condition to become true
+	ForCondition string `yaml:"forCondition" json:"forCondition,omitempty"`
+}
```
Comment (rsevilla87) [inline pkg/config/types.go:147]: This struct is not actually needed, you can define `WaitOptions` directly as string

File: pkg/burner/create.go [hunk 2]
```diff
@@ -141,6 +142,7 @@ func (ex *Executor) RunCreateJob(iterationStart, iterationEnd int) {
 		if ex.Config.PodWait {
 			if !ex.Config.NamespacedIterations || !namespacesWaited[ns] {
 				log.Infof("Waiting up to %s for actions to be completed in namespace %s", ex.Config.MaxWaitTimeout, ns)
+                                wg.Wait()
```
Comment (rsevilla87) [inline pkg/burner/create.go:145]:
```suggestion
				wg.Wait()
```

**Review by rsevilla87 — APPROVED** [review]
lgtm

**Pull request merged**